src/libraries/System.Net.Quic/src/System/Net/Quic/Implementations/MsQuic/MsQuicListener.cs

```diff
@@ -0,0 +1,182 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System.Net.Quic.Implementations.MsQuic.Internal;
+using System.Net.Security;
+using System.Runtime.InteropServices;
+using System.Threading;
+using System.Threading.Channels;
+using System.Threading.Tasks;
+using static System.Net.Quic.Implementations.MsQuic.Internal.MsQuicNativeMethods;
+
+namespace System.Net.Quic.Implementations.MsQuic
+{
+    internal class MsQuicListener : QuicListenerProvider, IDisposable, IAsyncDisposable
+    {
+        // Functions to invoke in MsQuic
+        private MsQuicApi _api;
+
+        // Security configuration for MsQuic
+        private MsQuicSecurityConfig _secConfig;
+
+        // Pointer to the underlying listener
+        private IntPtr _nativeObjPtr;
+
+        // Handle to this object for native callbacks.
+        private GCHandle _handle;
+
+        // Delegate that wraps the static function that will be called when receiving an event.
+        private ListenerCallbackDelegate _listenerDelegate;
+
+        // Ssl listening options (ALPN, cert, etc)
+        private SslServerAuthenticationOptions _sslOptions;
+
+        // To prevent multiple calls to StartAsync
+        private bool _started;
+
+        private volatile bool _disposed;
+
+        private readonly Channel<MsQuicConnection> _acceptConnectionQueue = Channel.CreateUnbounded<MsQuicConnection>(new UnboundedChannelOptions
+        {
+            SingleReader = true,
+            SingleWriter = true
+        });
+
+        internal MsQuicListener(IPEndPoint listenEndPoint, SslServerAuthenticationOptions sslServerAuthenticationOptions, MsQuicApi api, IntPtr nativeObjPtr)
+        {
+            _api = api;
+            _sslOptions = sslServerAuthenticationOptions;
+            ListenEndPoint = listenEndPoint;
+            _nativeObjPtr = nativeObjPtr;
+        }
+
+        internal override IPEndPoint ListenEndPoint { get; }
+
+        internal override async ValueTask<QuicConnectionProvider> AcceptConnectionAsync(CancellationToken cancellationToken = default)
```

Comment:
It looks like the connections can outpace our server and grow `_acceptConnectionQueue` infinitely. Can we implement a backlog configuration to avoid this?

Reply:
There are two ways you could implement a backlog on top of MsQuic. Call stop on the listener when you reach your limit, and then start it back up when you're ready for more; or just reject (return an error) new connections if you're over your limit.

Reply:
Implementing the backlog via the channel for now.